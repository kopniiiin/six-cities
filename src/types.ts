export enum OfferType {
  APARTMENT = `apartment`,
  HOTEL = `hotel`,
  HOUSE = `house`,
  ROOM = `room`
}

export enum City {
  PARIS = `Paris`,
  COLOGNE = `Cologne`,
  BRUSSELS = `Brussels`,
  AMSTERDAM = `Amsterdam`,
  HAMBURG = `Hamburg`,
  DUSSELDORF = `Dusseldorf`
}

export enum SortType {
  POPULAR = `popular`,
  TO_HIGH_PRICE = `price: low to high`,
  TO_LOW_PRICE = `price: high to low`,
  TOP_RATED = `top rated first`
}

export enum Path {
  MAIN = `/`,
  LOGIN = `/login`,
  OFFER = `/offer`,
  FAVORITES = `/favorites`
}

export enum ServerURL {
  LOGIN = `/login`,
  OFFERS = `/hotels`,
  FAVORITES = `/favorite`,
  REVIEWS = `/comments`
}

export enum ServerResponseStatus {
  OK = 200,
  UNAUTHORIZED = 401
}

export enum AuthorizationStatus {
  AUTHORIZED = `AUTHORIZED`,
  UNAUTHORIZED = `UNAUTHORIZED`
}

export interface Id {
  id: string;
}

export interface Location {
  coordinates: number[];
}

export interface User {
  name: string;
  photo: string;
}

export type Offer = Id & {
  type: OfferType;
  name: string;
  mainPhoto: string;
  isFavorite: boolean;
  isPremium: boolean;
  rating: number;
  price: number;
}

export interface Review {
  date: string;
  text: string;
  rating: number;
  user: User;
}

export type ReviewWithId = Id & Review;

export interface ReviewData {
  text: string;
  rating: number;
}

export interface AuthorizationData {
  email: string;
  password: string;
}
