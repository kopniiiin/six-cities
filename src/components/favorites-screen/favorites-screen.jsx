import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {Link} from "react-router-dom";

import {Path} from "../../const.js";

import OfferList from "../offer-list/offer-list.jsx";

import {Operation as OffersOperation} from "../../reducer/offers/offers.js";
import {getGroupedByCityFavoriteOffers} from "../../reducer/offers/selectors.js";

const propTypes = {
  children: PropTypes.element.isRequired,
  citiesToOffers: PropTypes.object.isRequired,
  loadData: PropTypes.func.isRequired,
  onOfferCardBookmarkButtonClick: PropTypes.func.isRequired
};

class FavoritesScreen extends PureComponent {
  render() {
    const {children, citiesToOffers, onOfferCardBookmarkButtonClick} = this.props;

    const offerGroups = Object.entries(citiesToOffers);

    const favoritesClassName = `page ${offerGroups.length ? `` : `page--favorites-empty`}`;
    const mainClassName = `page__main page__main--favorites ${offerGroups.length ? `` : `page__main--favorites-empty`}`;

    return (
      <div className={favoritesClassName}>
        {children}

        <main className={mainClassName}>
          <div className="page__favorites-container container">
            {offerGroups.length ? (
              <section className="favorites">
                <h1 className="favorites__title">Saved listing</h1>
                <ul className="favorites__list">
                  {offerGroups.map(([city, offers]) => (
                    <li key={city} className="favorites__locations-items">
                      <div className="favorites__locations locations locations--current">
                        <div className="locations__item">
                          <a className="locations__item-link">
                            <span>{city}</span>
                          </a>
                        </div>
                      </div>
                      {
                        <OfferList
                          blockClassName={`favorites`}
                          offers={offers}
                          onOfferCardMouseEnter={() => {}}
                          onOfferCardMouseLeave={() => {}}
                          onOfferCardBookmarkButtonClick={onOfferCardBookmarkButtonClick}/>
                      }
                    </li>
                  ))}
                </ul>
              </section>
            ) : (
              <section className="favorites favorites--empty">
                <h1 className="visually-hidden">Favorites (empty)</h1>
                <div className="favorites__status-wrapper">
                  <b className="favorites__status">Nothing yet saved.</b>
                  <p className="favorites__status-description">Save properties to narrow down search or plan yor future trips.</p>
                </div>
              </section>
            )}
          </div>
        </main>

        <footer className="footer container">
          <Link className="footer__logo-link" to={Path.MAIN}>
            <img className="footer__logo" src="img/logo.svg" alt="6 cities logo" width="64" height="33"/>
          </Link>
        </footer>
      </div>
    );
  }

  componentDidMount() {
    const {loadData} = this.props;

    loadData();
  }
}

FavoritesScreen.propTypes = propTypes;

const mapStateToProps = (state) => ({
  citiesToOffers: getGroupedByCityFavoriteOffers(state)
});

const mapDispatchToProps = (dispatch) => ({
  loadData: () => dispatch(OffersOperation.loadFavoriteOffers())
});

export {FavoritesScreen};
export default connect(mapStateToProps, mapDispatchToProps)(FavoritesScreen);
