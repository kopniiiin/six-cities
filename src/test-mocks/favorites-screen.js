import {OfferType} from "../const.js";

export default {
  citiesToOffers: {
    city: [{
      id: `4`,
      type: OfferType.ROOM,
      name: `Paper place`,
      mainPhoto: `photo`,
      isFavorite: false,
      isPremium: false,
      rating: 4,
      price: 4
    }]
  },
  loadData: () => {},
  onOfferCardBookmarkButtonClick: () => {}
};
