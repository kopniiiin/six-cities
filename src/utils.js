const MONTHS = [
  `January`,
  `February`,
  `March`,
  `April`,
  `May`,
  `June`,
  `July`,
  `August`,
  `September`,
  `October`,
  `November`,
  `December`
];

export const extend = (target, update) => Object.assign({}, target, update);

export const pushElement = (array, element) => [...array, element];

export const replaceElement = (array, element, index) => [...array.slice(0, index), element, ...array.slice(index + 1)];

export const removeElement = (array, index) => [...array.slice(0, index), ...array.slice(index + 1)];

export const upperCaseFirstLetter = (string) => string[0].toUpperCase() + string.slice(1);

export const formatDate = (dateString) => {
  const date = new Date(dateString);
  return `${MONTHS[date.getMonth()]} ${date.getDate()}, ${date.getFullYear()}`;
};
