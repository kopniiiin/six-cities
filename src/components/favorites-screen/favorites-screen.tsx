import * as React from "react";
import {connect} from "react-redux";
import {Link} from "react-router-dom";

import {Path, Offer} from "../../types";

import {doNothing} from "../../utils";

import OfferList from "../offer-list/offer-list";

import {Operation as OffersOperation} from "../../reducer/offers/offers";
import {getGroupedByCityFavoriteOffers} from "../../reducer/offers/selectors";

interface Props {
  children: React.ReactNode;
  citiesToOffers: {[city: string]: Offer[]};
  loadData: () => void;
  onOfferCardBookmarkButtonClick: (id: string) => void;
}

class FavoritesScreen extends React.PureComponent<Props> {
  props: Props;

  render() {
    const {children, citiesToOffers, onOfferCardBookmarkButtonClick} = this.props;

    const offerGroups = Object.entries(citiesToOffers);

    const favoritesClassName = `page ${offerGroups.length ? `` : `page--favorites-empty`}`;
    const mainClassName = `page__main page__main--favorites ${offerGroups.length ? `` : `page__main--favorites-empty`}`;

    return (
      <div className={favoritesClassName}>
        {children}

        <main className={mainClassName}>
          <div className="page__favorites-container container">
            {offerGroups.length ? (
              <section className="favorites">
                <h1 className="favorites__title">Saved listing</h1>
                <ul className="favorites__list">
                  {offerGroups.map(([city, offers]) => (
                    <li key={city} className="favorites__locations-items">
                      <div className="favorites__locations locations locations--current">
                        <div className="locations__item">
                          <a className="locations__item-link">
                            <span>{city}</span>
                          </a>
                        </div>
                      </div>
                      <OfferList
                        blockClassName={`favorites`}
                        offers={offers}
                        onOfferCardMouseEnter={doNothing}
                        onOfferCardMouseLeave={doNothing}
                        onOfferCardBookmarkButtonClick={onOfferCardBookmarkButtonClick}/>
                    </li>
                  ))}
                </ul>
              </section>
            ) : (
              <section className="favorites favorites--empty">
                <h1 className="visually-hidden">Favorites (empty)</h1>
                <div className="favorites__status-wrapper">
                  <b className="favorites__status">Nothing yet saved.</b>
                  <p className="favorites__status-description">Save properties to narrow down search or plan yor future trips.</p>
                </div>
              </section>
            )}
          </div>
        </main>

        <footer className="footer container">
          <Link className="footer__logo-link" to={Path.MAIN}>
            <img className="footer__logo" src="img/logo.svg" alt="6 cities logo" width="64" height="33"/>
          </Link>
        </footer>
      </div>
    );
  }

  componentDidMount() {
    const {loadData} = this.props;

    loadData();
  }
}

const mapStateToProps = (state) => ({
  citiesToOffers: getGroupedByCityFavoriteOffers(state)
});

const mapDispatchToProps = (dispatch) => ({
  loadData: () => dispatch(OffersOperation.loadFavoriteOffers())
});

export {FavoritesScreen};
export default connect(mapStateToProps, mapDispatchToProps)(FavoritesScreen);
