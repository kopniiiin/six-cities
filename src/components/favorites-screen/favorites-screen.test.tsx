import * as React from "react";
import {configure, shallow} from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";

import {OfferType, City} from "../../types";

import {doNothing, extend} from "../../utils";

import {FavoritesScreen} from "./favorites-screen";

configure({adapter: new Adapter()});

const testMocks = {
  citiesToOffers: {
    [City.AMSTERDAM]: [{
      id: `4`,
      type: OfferType.ROOM,
      name: `Paper place`,
      mainPhoto: `photo`,
      isFavorite: false,
      isPremium: false,
      rating: 4,
      price: 4
    }]
  },
  loadData: doNothing,
  onOfferCardBookmarkButtonClick: doNothing
};

describe(`snapshot test: FavoritesScreen component`, () => {
  it(`should render correctly`, () => expect(
      toJson(shallow(<FavoritesScreen {...testMocks}><div/></FavoritesScreen>))
  ).toMatchSnapshot());

  it(`should render message without offers`, () => expect(
      toJson(shallow(<FavoritesScreen {...extend(testMocks, {citiesToOffers: {}})}><div/></FavoritesScreen>))
  ).toMatchSnapshot());
});
