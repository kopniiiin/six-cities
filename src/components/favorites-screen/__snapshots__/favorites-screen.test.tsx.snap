// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test: FavoritesScreen component should render correctly 1`] = `
<div
  className="page "
>
  <div />
  <main
    className="page__main page__main--favorites "
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites"
      >
        <h1
          className="favorites__title"
        >
          Saved listing
        </h1>
        <ul
          className="favorites__list"
        >
          <li
            className="favorites__locations-items"
            key="Amsterdam"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                >
                  <span>
                    Amsterdam
                  </span>
                </a>
              </div>
            </div>
            <OfferList
              blockClassName="favorites"
              offers={
                Array [
                  Object {
                    "id": "4",
                    "isFavorite": false,
                    "isPremium": false,
                    "mainPhoto": "photo",
                    "name": "Paper place",
                    "price": 4,
                    "rating": 4,
                    "type": "room",
                  },
                ]
              }
              onOfferCardBookmarkButtonClick={[Function]}
              onOfferCardMouseEnter={[Function]}
              onOfferCardMouseLeave={[Function]}
            />
          </li>
        </ul>
      </section>
    </div>
  </main>
  <footer
    className="footer container"
  >
    <Link
      className="footer__logo-link"
      to="/"
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="img/logo.svg"
        width="64"
      />
    </Link>
  </footer>
</div>
`;

exports[`snapshot test: FavoritesScreen component should render message without offers 1`] = `
<div
  className="page page--favorites-empty"
>
  <div />
  <main
    className="page__main page__main--favorites page__main--favorites-empty"
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites favorites--empty"
      >
        <h1
          className="visually-hidden"
        >
          Favorites (empty)
        </h1>
        <div
          className="favorites__status-wrapper"
        >
          <b
            className="favorites__status"
          >
            Nothing yet saved.
          </b>
          <p
            className="favorites__status-description"
          >
            Save properties to narrow down search or plan yor future trips.
          </p>
        </div>
      </section>
    </div>
  </main>
  <footer
    className="footer container"
  >
    <Link
      className="footer__logo-link"
      to="/"
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="img/logo.svg"
        width="64"
      />
    </Link>
  </footer>
</div>
`;
